test/strings/basic.jl

```diff
@@ -428,6 +428,9 @@ foobaz(ch) = reinterpret(Char, typemax(UInt32))
 # issue #17624, missing getindex method for String
 @test "abc"[:] == "abc"
 
-# PR #18259 reverted because failing test below
-a = [x for x in String([0xcf, 0x83, 0x83, 0x83, 0x83])]
-@test a[1] == 'σ'
+# we assume for performance that next/nextind return values past the end of a
+# string's underlying data; this helps with performance of `done`.
```

Comment:
I think there is a mistaken assumption here that the previous implementation worked well for invalid UTF-8. It's only correct to say that it worked out for particular forms of invalid UTF-8; the previous implementation broke on other forms of invalid UTF-8 that this implementation works on.

```julia
julia> start(String([0x99]))
1

julia> done(String([0x99]), 1)
ERROR: BoundsError: attempt to access 1-element Array{UInt8,1}:
 0x99
  at index [0]
 in endof at ./unicode/utf8.jl:36
 in done at ./strings/basic.jl:36
```

I'm fine with adding a comment clarifying that only valid strings are supported, but it's not like this is a change in behaviour.